File: src/engine/numerals.rs
use deadpool_sqlite::Pool;
use serde_json::{json, Value};

const CASES: [&str; 7] = ["prathama", "dvitiya", "tritiya", "caturthi", "panchami", "sasthi", "saptami"];

pub async fn analyze_numeral(word: &str, pool: &Pool) -> Vec<Value> {
    let word_clone = word.to_string();
    pool.get().await.unwrap().interact(move |conn| {
        let mut results = Vec::new();
        let mut stmt = conn.prepare("SELECT * FROM numerals").unwrap();
        let mut rows = stmt.query([]).unwrap();

        while let Ok(Some(row)) = rows.next() {
            let base_form: String = row.get("base_form").unwrap_or_default();
            let gender: String = row.get("gender").unwrap_or_default();

            for case in CASES.iter() {
                let case_val: String = row.get(*case).unwrap_or_default();
                let forms: Vec<&str> = case_val.split(',').map(|s| s.trim()).collect();
                
                for (i, form) in forms.iter().enumerate() {
                    if form == &word_clone && !form.is_empty() {
                        let vacana = ["eka", "dvi", "bahu"][i];
                        results.push(json!({
                            "type": "numeral",
                            "base_form": base_form,
                            "gender": gender,
                            "case": case,
                            "vacana": vacana
                        }));
                    }
                }
            }
        }
        results
    }).await.unwrap()
}

File: engine/numerals.go
package engine

import (
	"database/sql"
	"strings"
)

var numeralCases = []string{
	"prathama", "dvitiya", "tritiya", "caturthi", "panchami", "sasthi", "saptami",
}

// AnalyzeNumeral looks up inflected numeral forms in the numerals table
func AnalyzeNumeral(db *sql.DB, word string) []map[string]any {
	var results []map[string]any

	rows, err := db.Query("SELECT * FROM numerals")
	if err != nil {
		return results
	}
	defer rows.Close()

	cols, err := rows.Columns()
	if err != nil {
		return results
	}

	vacanaList := []string{"eka", "dvi", "bahu"}

	for rows.Next() {
		columns := make([]any, len(cols))
		columnPointers := make([]any, len(cols))
		for i := range columns {
			columnPointers[i] = &columns[i]
		}

		if err := rows.Scan(columnPointers...); err != nil {
			continue
		}

		rowMap := make(map[string]string)
		for i, colName := range cols {
			if val, ok := columns[i].([]byte); ok {
				rowMap[colName] = string(val)
			} else if val, ok := columns[i].(string); ok {
				rowMap[colName] = val
			}
		}

		baseForm := rowMap["base_form"]
		gender := rowMap["gender"]

		for _, caseName := range numeralCases {
			caseVal := rowMap[caseName]
			if caseVal == "" {
				continue
			}

			forms := strings.Split(caseVal, ",")
			for i, form := range forms {
				form = strings.TrimSpace(form)
				if form == word && form != "" && i < len(vacanaList) {
					results = append(results, map[string]any{
						"type":      "numeral",
						"base_form": baseForm,
						"gender":    gender,
						"case":      caseName,
						"vacana":    vacanaList[i],
					})
				}
			}
		}
	}

	return results
}